Fix Output:Table:SummaryReports monthly reports and Variable Name case-sensitivity in Output:Variable

jmarrec opened the pull request:
Pull request overview
---------------------

Fix #7019 Output:Table:SummaryReports doesn't work without variables.
Fix case-sensitivity in Output:Variable "Variable Name".

The main issue what that the `DataOutputs::FindItemInVariableList` that fails with different casing, so I made that all upper case.

I had originally actually requested the output variables needed for the report (so they show in SQL), but it isn't needed for the table to work correctly so I reverted this.

Added a unit test that fails without the fix and passes with it.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [x] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

jmarrec committed:
Add unit test for #7019

jmarrec committed:
Failed attempt at actually requesting the variables

jmarrec committed:
Make DataOutputs::OutputVariablesForSimulation store as upper case always, and saerch for upper case too

jmarrec committed:
Still not a good unit test...

jmarrec committed:
Revert of "Failed attempt at actually requesting the variables"

jmarrec committed:
Unit test now passes

jmarrec commented:
https://github.com/jmarrec/EnergyPlus/commit/b99e1c9dabb4a3b6c779f1f143da06f8c6cf336e is develop + unit test only. When running it it fails as expected.

```
[ RUN      ] EnergyPlusFixture.OutputReportTabularMonthlyPredefined_FindNeededOutputVars
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/OutputReportTabular.unit.cc:6695: Failure
      Expected: OutputReportTabular::MonthlyInput(1).numTables
      Which is: 0
To be equal to: 2
```

Also, I checked that that user defect file does work with it (and that I get actual, non zero unmet hours). I also created a simpler example : [1ZoneUncontrolled.idf.txt](https://github.com/NREL/EnergyPlus/files/2511341/1ZoneUncontrolled.idf.txt)

Without the fix, the HTML report is completely empty save for the header. With it, it works.

jmarrec commented:
I’ll need to fix the unit tests that do fail regarding finditemInVariableList. Will do that tomorrow

jmarrec committed:
Make DataOutputs::OutputVariablesForSimulation a case-insentive unordered_map

Implemented a insensitive hasher and comparator

jmarrec committed:
Update InputProcessor to use that case insensitive unordered_map

jmarrec committed:
OutputProcessor update: no need to make upper before finditemInVariableList

jmarrec committed:
Small extension of an existing unit test that would prove the case-insensitive behavior

jmarrec commented:
Made DataOutputs::OutputVariableForSimulations a nested case-insensitive std::unordered_map by implementing case-insentive hasher and comparator. All tests pass now including one I (barely) extended that would check this new case-insentive behavior.

mbadams5 commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -5550,7 +5547,9 @@ void SetupOutputVariable(std::string const &VariableName,           // String Na
-    VarName = VariableName;
+    // Variable name without units
+    std::string VarName = VariableName;
+    //  CHARACTER(len=MaxNameLength) :: VariableNamewithUnits ! Variable name with units std format
```
You don't need this line.

```suggestion
```

jmarrec replied:
Just an FYI, I didn't add this comment, I just move the varName declaration (and the comment) to initialize on the spot rather than at the top of the file.

mbadams5 commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -5785,7 +5783,9 @@ void SetupOutputVariable(std::string const &VariableName,           // String Na
-    VarName = VariableName;
+    // Variable name without units
+    std::string VarName = VariableName;
+    //  CHARACTER(len=MaxNameLength) :: VariableNamewithUnits ! Variable name with units std format
```
You don't need this line.

```suggestion
```

mbadams5 commented on src/EnergyPlus/DataOutputs.hh:
```diff
@@ -94,14 +95,41 @@ namespace DataOutputs {
+
+    // Two structs to have a case insensitive unordered_map. We need to have a case insenstive hasher and a case insensitive comparator
+    // (The default allocator for unordered_map is fine)
+    struct case_insensitive_hasher
```
These could potentially go to `UtilityRoutines.hh` since there could be other containers that could use these. However, this isn't critical since they are fine here as well.

mbadams5 reviewed (approved):
Looks good to me. I reviewed for potential impact to the regex Output:Variable usage but these changes don't appear to affect that.

jmarrec committed:
Delete useless comments (review)

jmarrec committed:
Move the case insensitive comparator and hasher into UtilityRoutines

Move the actual operator implementation into the .cc so we can change the implementation without having to recompile files referencing the header later.

jmarrec commented:
Thanks for the review @mbadams5. I deleted the two comments and move the comparators into UtilityRoutines (I put the actual implementation of the operators into the .cc file so we can change it without having to recompile all files referencing the UtilityRoutines.hh too)

I recompiled and ran the tests, they still pass just fine (though I know the bots will check that too)

JasonGlazer commented:
@jmarrec this seems to fix all the issues that I was having with #6863

nrel-bot-3 commented:
@jmarrec @lgentile it has been 14 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 7019_MissingVariableForReport

mjwitte reviewed (approved):
Tested with defect file, ran unit tests, all OK. Updated to current develop and will merge once CI reports back.

mjwitte commented:
Looking at the CI diffs, noticed that 5ZoneWaterCooled_GasFiredSteamHumidifier.idf is showing a csv diff (but only on the detailed regression results page):
```
Not comparing field MAIN GAS HUMIDIFIER:Humidifier Gas Use Rate [W](Hourly), which appears in input files . . . 
```
Turns out that this test file has one `Output:Variable` object with a stray uppercase letter in the output variable name
```
  Output:Variable,*,Humidifier Gas USe Rate,hourly;
```
This variable has been missing from the eso and csv output. Now it's there. Just to confirm that `Output:Variable` "Variable Name" has been case-sensitive (probably since ???), pushed the entire output variable section of this idf to uppercase and only two variables showed up in the csv output using v9.0.1 - the two that are metered. With this branch, all of the variables appear (same as the non-uppercase idf). Title and description for this PR will be updated to include fixing case-sensitive variable names.

@Myoldmopar It's disappointing that none of these significant diffs (additional csv variable, table diffs, and additional tables) show in the CI summary comments. For this file, it's just a warning with
```
[decent_ci:test_result:message] AUD diffs.
[decent_ci:test_result:message] MTD diffs.
```
Will post some new issues in [regression tool](https://github.com/NREL/EnergyPlusRegressionTool/issues) to better flag these diffs so CI can report them.
